### PR TITLE
Fix tstz DateTime normalization

### PR DESCRIPTION
**Pull request by MaceWindu**
Fix https://github.com/linq2db/linq2db.EntityFrameworkCore/issues/213
Fix https://github.com/linq2db/linq2db.EntityFrameworkCore/issues/316
Fix https://github.com/linq2db/linq2db/issues/3895
Fix https://github.com/linq2db/linq2db/issues/3776

### CODE REVIEW

**Comment by MaceWindu** [comment]
my bad

**Review by viceroypenguin — APPROVED** [review]

File: Source/LinqToDB/DataProvider/PostgreSQL/PostgreSQLDataProvider.cs [hunk 1]
```diff
@@ -231,14 +231,18 @@ public override SchemaProvider.ISchemaProvider GetSchemaProvider()
...
-					if (dt.Kind == DateTimeKind.Utc
-						&& (dataType.DataType == DataType.DateTime2 || GetNativeType(dataType.DbType) == NpgsqlProviderAdapter.NpgsqlDbType.Timestamp))
-						value = DateTime.SpecifyKind(dt, DateTimeKind.Unspecified);
 					// timestamptz should have UTC Kind
-					else if (dt.Kind != DateTimeKind.Utc
-						&& (dataType.DataType == DataType.DateTimeOffset || GetNativeType(dataType.DbType) == NpgsqlProviderAdapter.NpgsqlDbType.TimestampTZ))
-						value = DateTime.SpecifyKind(dt, DateTimeKind.Utc);
+					if ((dataType.DataType == DataType.DateTimeOffset || GetNativeType(dataType.DbType) == NpgsqlProviderAdapter.NpgsqlDbType.TimestampTZ))
+					{
+						if (dt.Kind != DateTimeKind.Utc)
```
[the remark this answers is not in the document]
Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/PostgreSQL/PostgreSQLDataProvider.cs:237]: :man_shrugging: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L1331-L1335

[the remark this answers is not in the document]
Comment (viceroypenguin) [reply Source/LinqToDB/DataProvider/PostgreSQL/PostgreSQLDataProvider.cs:237]: 🤦🏼‍♂️ fair enough. LGTM as is then.

**Comment by Shane32** [comment]
Correct me if I'm wrong, but the code appears to have exactly identical logic as previously.

**Comment by MaceWindu** [comment]
Previous logic identified column with `DataType.DateTime2` (default value for `DateTime`) as `timestamp` even if `DbType` set to `tstz` type

**Comment by Shane32** [comment]
It's refactoring, but will behave identical.

Consider if the secondary `if` (e.g. `dt.Kind == DateTimeKind.Utc`) are rolled into the if above.  While this is a logic change, there is no conflict because they match on different `dt.Kind` values.  Then compare the result to previously.  They are simply in a differnent order.  Since `dt.Kind` is used match, it is identical behavior.

**Comment by Shane32** [comment]
I don't mind the refactoring of course, but I just want to point out that it can't possibly fix any bugs if the logic hasn't changed.

**Comment by viceroypenguin** [comment]
> It's refactoring, but will behave identical.
> 
> Consider if the secondary `if` (e.g. `dt.Kind == DateTimeKind.Utc`) are rolled into the if above. While this is a logic change, there is no conflict because they match on different `dt.Kind` values. Then compare the result to previously. They are simply in a differnent order. Since `dt.Kind` is used match, it is identical behavior.

There is a difference. Consider the following simplified pseudo code:
```csharp
var bool1 = true;
var bool2 = false;

// ex1: `else` only requires one fo them to be false.
if (bool2 && bool1)
{
  // both must be true
}
else
{
  // either one is false
}

// ex2: `else` only requires that bool1 is false
if (bool1)
{
  if (bool2)
  {
    // both must be true
  }
}
else
{
  // bool1 is false, irrespective of bool2
}
```

**Comment by Shane32** [comment]
Yes, but we have this:

```cs
if (bool1)
{
  if (bool2)
  {
    // both must be true
  }
}
else
{
  if (!bool2)
  {
    // bool1 and bool2 must both be false
  }
}
```

So it can be simplified to this:

```cs
if (bool1 && bool2)
{
  // both must be true
}
else if (!bool2)
{
  // bool1 and bool2 must both be false
}
```

At which point it becomes identical to the previous code.

**Comment by Shane32** [comment]
The secondary if conditions are mutually exclusive

image

**Comment by MaceWindu** [comment]
inputs:
- DateTime with Kind = UTC
- DataType: DateTime2
- DbType: "tstz"

Old code fails to handle it correctly: it executes first `if` treating value as `ts without time zone` and converting it to `Unspecified` kind

New code:
- identify it as `tstz` by using `DbType` check
- doesn't try to execute `else` branch when `tstz` type identified but no kind change required

**Comment by Shane32** [comment]
Ok, I think I'm understanding.  Due to the `||` operation in the first if, the conditions within the `else` may partially match, so the order of the `if` condition matters.  This also requires the secondary if be nested.  Got it - thanks!!

**Comment by Shane32** [comment]
E.g. with the `||` in play, we can have this situation

```cs
var bool1 = true;
var bool3 = true;

if (bool1) {
   if (bool2) ...
} else if (bool3) {
   if (!bool2) ...
}
```

So when bool2 == false, no code should execute.

**Review by igor-tkachev — APPROVED** [review]

**Pull request merged**